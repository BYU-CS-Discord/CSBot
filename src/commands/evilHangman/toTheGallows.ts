import { EmbedBuilder, SlashCommandBuilder } from 'discord.js';
import { appVersion } from '../../constants/meta';
import type { GameStore } from '../../evilHangman/gameStore';

const builder = new SlashCommandBuilder()
	.setName('tothegallows')
	.setDescription('Begins a new game of Evil Hangman');

export function toTheGallows(gameStore: GameStore): GlobalCommand {
	return {
		info: builder,
		requiresGuild: false,
		async execute({ reply, channelId }): Promise<void> {
			const generalEmbed = new EmbedBuilder()
				.setTitle('Evil Hangman')
				.setFooter({ text: `v${appVersion}` });

			if (gameStore.games.has(channelId)) {
				const embed = generalEmbed
					.setDescription('There is already a game running in this channel')
					.setColor('DarkRed');

				await reply({
					embeds: [embed],
					ephemeral: true,
				});
			} else {
				const embed = generalEmbed.setDescription('Placeholder');
				gameStore.games.set(channelId, 0); // TODO: 0 is a placeholder

				await reply({
					embeds: [embed],
				});
			}
		},
	};
}
