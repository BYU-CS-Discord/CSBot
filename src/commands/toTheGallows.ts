import { SlashCommandBuilder } from 'discord.js';
import { buildEvilHangmanMessage } from '../evilHangman/evilHangmanMessage.js';
import { EvilHangmanGame } from '../evilHangman/evilHangmanGame.js';

const LengthOption = 'wordlength';
const GuessesOption = 'numguesses';

const builder = new SlashCommandBuilder()
	.setName('tothegallows')
	.setDescription('Begins a new game of Evil Hangman')
	.addIntegerOption(option =>
		option
			.setName(LengthOption)
			.setDescription('The number of letters in the word to guess')
			.setMinValue(1)
	)
	.addIntegerOption(option =>
		option
			.setName(GuessesOption)
			.setDescription('The number of allowed incorrect guesses')
			.setMinValue(1)
			.setMaxValue(25)
	);

export const toTheGallows: GlobalCommand = {
	info: builder,
	requiresGuild: false,
	async execute({ reply, options }): Promise<void> {
		const wordLength = options.getInteger(LengthOption);
		const numGuesses = options.getInteger(GuessesOption);

		const game = EvilHangmanGame.newGame(wordLength, numGuesses);
		const response = await buildEvilHangmanMessage(game.getDisplayInfo());

		await reply(response);
	},
};
