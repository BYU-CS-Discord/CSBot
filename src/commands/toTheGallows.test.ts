import { afterEach, assert, beforeEach, describe, expect, test, vi } from 'vitest';

import { toTheGallows } from './toTheGallows.js';

vi.mock('../constants/meta.js', async () => {
	const { repo } =
		await vi.importActual<typeof import('../constants/meta.js')>('../constants/meta.js');
	return {
		// Version changes frequently, so use a consistent version number to test with:
		appVersion: 'X.X.X',
		repo,
	};
});

describe('toTheGallows', () => {
	const mockReply = vi.fn<TextInputCommandContext['reply']>();
	const mockGetInteger = vi.fn<TextInputCommandContext['options']['getInteger']>();
	let context: TextInputCommandContext;

	beforeEach(() => {
		context = {
			reply: mockReply,
			options: {
				getInteger: mockGetInteger,
			},
		} as unknown as TextInputCommandContext;

		mockGetInteger.mockReturnValue(null);
	});

	afterEach(() => {
		vi.resetAllMocks();
	});

	test('begins a game of evil hangman', async () => {
		await expect(toTheGallows.execute(context)).resolves.toBeUndefined();

		expect(mockReply).toHaveBeenCalledOnce();
		const response = mockReply.mock.calls.at(0)?.at(0);
		if (!response || typeof response === 'string' || !('embeds' in response)) {
			assert.fail('Did not reply to interactions with embeds');
		}
		expect(response.embeds.length).toEqual(1);
		expect(response.embeds.at(0)).toBeTypeOf('object');
		expect(response.components?.length).toEqual(5);
	});

	test('specifying length and number of guesses always puts the game into the same state', async () => {
		mockGetInteger.mockReturnValue(4);
		await expect(toTheGallows.execute(context)).resolves.toBeUndefined();

		expect(mockReply).toHaveBeenCalledOnce();

		const response = mockReply.mock.calls.at(0)?.at(0);
		if (!response || typeof response === 'string' || !('embeds' in response)) {
			assert.fail('Did not reply to interactions with embeds');
		}
		expect(response.embeds.length).toEqual(1);
		const embed = response.embeds.at(0);
		const embedData = 'toJSON' in embed ? embed.toJSON() : embed;
		expect(embedData).toMatchSnapshot();
		expect(response.components?.length).toEqual(5);
	});
});
