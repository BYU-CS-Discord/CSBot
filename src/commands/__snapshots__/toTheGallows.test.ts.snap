// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toTheGallows specifying length and number of guesses always puts the game into the same state 1`] = `
{
  "fields": [
    {
      "inline": true,
      "name": "_ _",
      "value": "\`\`\` /---\\
     |
     |
     |
     |
    -|-
\`\`\`",
    },
    {
      "inline": true,
      "name": "_ _",
      "value": "Remaining Guesses: 4
Word: ----
Letters Guessed: _ _",
    },
  ],
  "footer": {
    "icon_url": undefined,
    "text": "vX.X.X",
  },
  "title": "Evil Hangman",
}
`;
